### ISSUE TITLE
U/Int64 conversion 

### ISSUE THREAD

**Issue by sirudog** [gopherjs/gopherjs#360]
Hi,
Recently I came across an issue of javascript not being able to handle numbers larger than 2^53 - 1. 
This is because javascript **number** type is specified as an IEEE754 floating-point value, so it is not guaranteed to be able to hold a 64-bit unsigned value. Floating-point values can only represent values up to 2^53 - 1 without losing precision.
So I had to use a js library to use 64 bit integers correctly and safely. (there are several libs for this out there by the way, I chose bignumber.js (https://github.com/MikeMcl/bignumber.js))

Based on these definitions found in go.js (https://github.com/gopherjs/gopherjs/blob/master/js/js.go)
```javascript
// Int64 returns the object converted to int64 according to JavaScript type conversions (parseInt).
func (o *Object) Int64() int64 { return o.object.Int64() }

// Uint64 returns the object converted to uint64 according to JavaScript type conversions (parseInt).
func (o *Object) Uint64() uint64 { return o.object.Uint64() }
```
it is not clear if I need to take care of this, or it is being taken care of by gopherjs, and if so, how this conversion is being done, and what would be the proper way to handle these following scenarios:
1. JS => Go 
function in Go:
```func TakeMyInt64(myInt64 uint64) { ... }```
when I call it from JS, I may not pass an integer of type **number** directly to this function, as I have to use a custom uint64 object from a JS library to handle 64 bit ints correctly. 
2. Go => JS
function in Go:
```func GetInt64() uint64 { ... }```
when I call it from JS, what will I get back? Does gopherjs use its own internal type object to represent 64 bit integers? Is this object what I have found declared in the generated JS code?
```javascript
var $Uint64        = $newType( 8, $kindUint64,        "uint64",         "uint64",     "", null);
```
It looks like I will have to do a conversion between $Uint64 (GoJS) and BigNumber (bignumber.js lib). 
Any help/opinion is much appreciated

**Comment by neelance** [comment]
You are right, JS can not express more than 53 bit integers without loosing precision.

The easiest solution would be to do all your calculations in Go code, since GopherJS has full support for 64 bit integers.

If you really need to also use JS code with bignumber.js besides Go code, then I'd suggest to write Go helper functions that convert `int64`/`uint64` to their bignumber.js equivalents by splitting the integer into two 32 bit integers via shifts and casts. Then pass those two to JS and do the reverse shift and addition to put them back together into a bignumber.js value.

**Comment by sirudog** [comment]
Thanks for the hints! I managed to handle JS => Go case by changing the function which previously accepted an uint64 param. Now it accepts a string param (so that I can pass bignumber.js number.toString()), and I do the conversion in go like this:
```
value, err := strconv.ParseUint(strValue, 10, 64)
```
But the Go => JS case is a bit more difficult, as I do not directly return an uint64 from my Go func. If I did, I would be able to do: 
```
return strconv.FormatUint(uint64Value, 10)
```
so that in JS by receiving the string representation of uint64 I could create a bignumber.js number. 
What I want to return is a struct (which has an uint64 field) that I wrap with a MakeWrapper call:
```
func GetLogonRequestFrom(buffer []byte) *js.Object {
    // some code to turn buffer into a LogonRequest struct, which has an uint64 field
    return js.MakeWrapper(logonRequest)
}
```
so when I call this function from js:
```
var logonRequest = backendMessages.GetLogonRequestFrom(logonRequestBytes);
var myUint64Value = logonRequest.Uint64Value(); => precision loss
```
I see a precision loss during externalizing this field value, caused by:
```
var $flatten64 = function(x) {
  return x.$high * 4294967296 + x.$low;
};
```
which is used by $externalize function. 
I see that the value of this uint64 typed field is correctly handled as two uint32 vars internally (x param), but when it is being externalized, $flatten64 is called on it, which just converts it to a number. I think this function is not "safe" to use for externalizing u/int64 values. 
I also see that the same $flatten64 function is used for Dates. Does it mean that Dates are also not safe to use in return values of my Go functions?

What I am trying to say is that this behaviour at least should be documented better with a warning about possible precision losses, and what would be better is to return the string representation of the uint64 value from the $flatten64 function. Most of the javascript libraries dealing with int64 values can accept strings. (sometimes even the high and low 32 bit integers)

Now maybe I am completely off-track with this, and I know that I can do a workaround by not directly returning my struct which has uint64 fields, or in general any data structure having uint64 members, but  having an extra layer of wrapper objects just to make the uint64=>string conversion in Go, before I call js.MakeWrapper on them does not seem to be the right way and does not scale at all.
I am curious what do you think about this flattening issue, and if there is anything gopherjs could do better/differently, or is it out of scope of how gopherjs generates JS code?
